### PR TITLE
Add all-projects support for network zones #712

### PR DESCRIPTION
**Pull request by milaiwi**
image

image

### What Changed

1. api: network_zones_all_projects

-  Added `network_zones_all_projects` to `internal/version/api.go`

2. incusd/network_zones: Add support for all-projects

- Added support to all-projects by getting all network zones when allProjects field is added to the uri in `networkZonesGet` function.
- Included all-projects to the comments for `networkZonesGet` functon
- Added Project field in the `NetworkZone` struct in `shared/api/network_zone.go`

3. client: Add GetNetworkZonesAllProjects
- Added `GetNetworkZonesAllProjects` that makes a GET request with all-projects field. Added it to `client/interfaces.go`

4. incus/network_zone: Add --all-projects flag to list
- Added flagAllProjects boolean field to cmdNetworkZoneList
- Calls GetNetworkZonesAllProjects when set to true, GetNetworkZones on false
- Changed output to include Project in the table when set

### CODE REVIEW

File: client/incus_network_zones.go [hunk 1]
```diff
@@ -42,6 +42,22 @@ func (r *ProtocolIncus) GetNetworkZones() ([]api.NetworkZone, error) {
 	return zones, nil
 }
 
+// GetNetworkZonesAllProjects returns a list of network zones across all projects as NetworkZone structs
+func (r *ProtocolIncus) GetNetworkZonesAllProjects() ([]api.NetworkZone, error) {
+    err := r.CheckExtension("network_dns")
```
Comment (stgraber) [inline client/incus_network_zones.go:47]: Should be `network_zones_all_projects` instead

File: client/incus_network_zones.go [hunk 2]
```diff
@@ -42,6 +42,22 @@ func (r *ProtocolIncus) GetNetworkZones() ([]api.NetworkZone, error) {
...
+    if err != nil {
+		return nil, fmt.Errorf(`The server is missing the required "network_dns" API extension`)
```
Comment (stgraber) [inline client/incus_network_zones.go:49]: Here too

File: cmd/incus/network_zone.go [hunk 3]
```diff
@@ -122,15 +124,25 @@ func (c *cmdNetworkZoneList) Run(cmd *cobra.Command, args []string) error {
...
+            return err
+        }
+    }
+
 
 	data := [][]string{}
 	for _, zone := range zones {
 		strUsedBy := fmt.Sprintf("%d", len(zone.UsedBy))
 		details := []string{
+            zone.Project,
```
Comment (stgraber) [inline cmd/incus/network_zone.go:145]: Similar to what we've done elsewhere, the extra Project column should only be showed if `flagAllProjects` is true

File: cmd/incusd/network_zones.go [hunk 4]
```diff
@@ -143,11 +149,34 @@ func networkZonesGet(d *Daemon, r *http.Request) response.Response {
...
 
-	var zoneNames []string
+	// var zoneNames []string
+    var zoneNamesMap map[string]string
+    allProjects := util.IsTrue(request.QueryParam(r, "all-projects"))
 
 	err = s.DB.Cluster.Transaction(r.Context(), func(ctx context.Context, tx *db.ClusterTx) error {
 		// Get list of Network zones.
-		zoneNames, err = tx.GetNetworkZonesByProject(ctx, projectName)
+		//zoneNames, err = tx.GetNetworkZonesByProject(ctx, projectName)
```
Comment (stgraber) [inline cmd/incusd/network_zones.go:158]: That line should probably just go away instead of being commented.

**Comment by stgraber** [comment]
Just left a few notes on small things to clean up, but overall that looks good.

You're still going to need to add a commit `i18n: Update translation templates` whose content you can generate by running `make i18n`.

To make the doc test happy, you'll need to update `doc/api-extensions.md` and remove some of those blank lines around line 1849/1854.

File: doc/api-extensions.md [hunk 5]
```diff
@@ -1846,6 +1846,12 @@ This adds:
 * `PATCH /1.0/network-zones/ZONE/records/RECORD`
 * `DELETE /1.0/network-zones/ZONE/records/RECORD`
 
+
```
Comment (stgraber) [inline doc/api-extensions.md:1849]: Remove the line

File: doc/api-extensions.md [hunk 6]
```diff
@@ -1846,6 +1846,12 @@ This adds:
...
+## `network_zones_all_projects`
+
+This adds support for listing network zones across all projects through the `all-projects` parameter on the `GET /1.0/network-zones`API.
+
+
```
Comment (stgraber) [inline doc/api-extensions.md:1854]: Remove the line

File: client/incus_network_zones.go [hunk 7]
```diff
@@ -42,6 +42,22 @@ func (r *ProtocolIncus) GetNetworkZones() ([]api.NetworkZone, error) {
...
+    if err != nil {
+		return nil, fmt.Errorf(`The server is missing the required "network_zones_all_projects" API extension`)
+    }
+
+    zones := []api.NetworkZone{}
+    uri := "/network-zones?all-projects=true&recursion=1"
+    _, err = r.queryStruct("GET", uri, nil, "", &zones)
+    if err != nil {
+        return nil, err
+    }
```
Comment (stgraber) [inline client/incus_network_zones.go:57]: Add a line break here

File: shared/api/network_zone.go [hunk 8]
```diff
@@ -44,6 +44,13 @@ type NetworkZone struct {
 	// Read only: true
 	// Example: ["/1.0/networks/foo", "/1.0/networks/bar"]
 	UsedBy []string `json:"used_by" yaml:"used_by"` // Resources that use the zone.
+
```
Comment (stgraber) [inline shared/api/network_zone.go:47]: Remove the line break

File: internal/version/api.go [hunk 9]
```diff
@@ -408,7 +408,7 @@ var APIExtensions = []string{
 	"network_integrations",
 	"instance_memory_swap_bytes",
 	"network_bridge_external_create",
-    "network_zones_all_projects"
+    "network_zones_all_projects",
```
Comment (stgraber) [inline internal/version/api.go]: This change should be merged into the `api: XYZ` commit

**Comment by stgraber** [comment]
Looking great, just a few minor nits I commented in-line!

On top of those, it'd be good to add another commit:
 - `tests: Add all-projects network zone test`

I'd keep that very simple, just edit `test/suites/network_zone.sh` and just after the `Create zone in project` part, make a quick test of `incus network zone list --all-projects` to ensure that you see both `incus-foo.example.net` (foo project) and `incus.example.net` (default project).

**Comment by stgraber** [comment]
And then this will be ready to merge, thanks!

**Comment by milaiwi** [comment]
For the test suite, what should the behavior on a failure be? Additionally, since the check would have to look at the content of the table, would something like this be appropriate?

`output=$(incus network zone list --all-projects)

if output=$(incus network zone list --all-projects) && \
   echo "$output" | grep -q "incus-foo.example.net" && \
   echo "$output" | grep -q "incus.example.net"; then
    exit 0
else
    exit 1
fi`

**Comment by stgraber** [comment]
I'd probably do something like:
```
incus network zone list --all-projects -fcsv | grep -q default,incus.example.net || false
incus network zone list --all-projects -fcsv | grep -q foo,incus-foo.example.net || false
```

So that uses the csv output which should be easier to check and then validates we get both the default and foo project entries.

**Comment by stgraber** [comment]
Oh, I missed that you added the test, in the future, please leave a comment to make sure I get notified!

Going to give that one a quick rebase now and see if we can't get it merged!

**Comment by stgraber** [comment]
Did a few fixes on this one:
 - Rebased it to resolve the conflict
 - Ran `gofmt -s -w ./` and applied the fixes to the various commits
 - Ran `make static-analysis` to look for anything else that needed fixing
 - Picked up bad comment syntax in client/incus_network_zones.go (missing period)
 - Picked up indentation that could be optimized in cmd/incusd/network_zones.go
 - Fixed both issues and applied changes to respective commits
 - Ran `make static-analysis` again to check for anything else
 - It reported that `make update-api` needed to be run, so ran that and committed the result
 - Ran `make static-analysis` again to see if all clean now
 - All good so proceeded to a manual commit review
 - Commit `api: network_zones_all_projects` incorrect modifies `shared/api`, moved that bit into the applicable `shared/api` commit instead.
 - Commit `incusd/network_zones: Add support for all-projects` somehow seriously mangled the swagger API description comment (tab/space issue), manually fixed it

Now pushed the result so just need to see if I missed anything :)

**Review by stgraber — APPROVED** [review]

**Pull request merged**